moveit_core/collision_detection/src/world.cpp

```diff
@@ -136,6 +136,91 @@ bool World::hasObject(const std::string& object_id) const
   return objects_.find(object_id) != objects_.end();
 }
 
+bool World::knowsTransform(const std::string& object_id) const
+{
+  // Check object names first
+  if (objects_.find(object_id) != objects_.end())
+    return true;
+  else  // Then objects' subframes
+  {
+    for (const std::pair<std::string, ObjectPtr>& object : objects_)
+    {
+      try 
+      {
+        // Strip the object's name that should be prepended to the requested frame
+        int name_length = object.second->id_.length();
+        if ("/" == object_id.substr(name_length, name_length+1))
+          if (object.second->id_ == object_id.substr(0,name_length))
+          {
+            if (object.second->subframe_poses_.canTransform(object_id.substr(name_length+1)))
+              return true;
+          }
+      }
+      catch (const std::exception& e) {;}
+    }
+  }
+  return false;
+}
+
+const Eigen::Isometry3d& World::getTransform(const std::string& object_id) const
+{
+  // This is almost identical to the function below, except for the return.
+  std::map<std::string, ObjectPtr>::const_iterator it = objects_.find(object_id);
+  if (it != objects_.end())
+    return it->second->shape_poses_[0];
+  else  // Find within subframes
+  {
+    for (const std::pair<std::string, ObjectPtr>& object : objects_)
```

Comment:
I really prefer calls of the form `getTransform(object_id, subframe_id)` because they avoid all this for loop stuff. These calls in the world class are used by people in inner loops of algorithms on robots. In this implementation, some calls will look through every child frame of every object before returning, even if they have nothing to do with that object. I think that is way to high a price to pay to get the use-friendly one-string-with-slashes naming format. @BryceStevenWilley @rhaschke do you disagree?

Reply:
I don't disagree, but I think that style of search could be done with the current interface, as long as collision object names don't have slashes. Just search for a slash, if it's there, everything before is the object name and everything after is the subframe.  Not sure what restrictions there are on collision object names, but I feel like saying they shouldn't include slashes isn't too strict. 

I also had the impression that there wouldn't be an exorbitant amount of subframes in the whole scene, maybe 2 or 3 for some objects of interest. I could be wrong there though.

Reply:
(I realized after i said this that string comparison is linear time anyway) but still don't like the idea of creating a parsed naming scheme within the planning scheme that looks like tf but isn't.

Reply:
I can understand the concern, but I can't evaluate the performance impact. If we ask for slashes not to be in the object name, it would only be one linear `string::find` operation for each call to `knowsTransform`. `getTransform` is not affected if no subframes are used. Not using slashes in names is so common that I don't think it would even be perceived as a restriction. I moved the check for a slash separator before the for-loop in world and robot_state for now, so this can move one way or another.

Second, I understand that the concern is about the nested lookups instead of through a dedicated additional function. @rhaschke proposed (in our call, not yet in an issue) to combine the robot state and collision world into one class with a TF-like hierarchical TransformMap for all fixed transforms. That way, each subframe would just become another entry in the map, so the search through them would not be as heavy. The current unified interface would be easy to update behind the scenes.

Last, I don't think that the interface will be confusing to users. In fact, "looks and works sort of like TF" might well make it easier to understand, because they already know TF. But that's just the UI end of things, naturally.

Reply:
I agree with what felixvd and rhaschke are proposing, that'd definitely get around the performance issue, and would go a long way to resolving https://github.com/ros-planning/moveit/issues/202.

While I do agree that looks like TF does make it easier to understand, I think what @jonbinney is saying is that users might *think* it's actually TF, mistakenly trying to publish subframes through TF instead of going through the move group interface? Am I correct in that interpretation?

Reply:
Yes, that they might think it is TF, or that since the frames have slashes like TF, that other TF-like behavior might be supported.   

My issue with this was purely a preference though, and I'm convinced just by hearing that others are comfortable with it. Lets move forward with this PR.